spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -143,36 +148,45 @@ public Subscriber subscribe(String subscription, MessageReceiver messageHandler)
 	/**
 	 * Pulls messages synchronously, on demand, using the pull request in argument.
 	 *
-	 * <p>This method acknowledges all received messages.
 	 * @param pullRequest pull request containing the subscription name
 	 * @return the list of {@link PubsubMessage} containing the headers and payload
 	 */
-	private List<PubsubMessage> pull(PullRequest pullRequest, RetrySettings retrySettings) {
+	private List<PubsubMessage> pull(PullRequest pullRequest, RetrySettings retrySettings,
+			boolean acknowledge) {
 		Assert.notNull(pullRequest, "The pull request cannot be null.");
+		List<PubsubMessage> pulledMessages = new ArrayList<>();
 
 		try {
 			SubscriberStub subscriber = this.subscriberFactory.createSubscriberStub(retrySettings);
 			Assert.notNull(subscriber, "A SubscriberStub is needed to execute the pull request.");
 
 			PullResponse pullResponse =	subscriber.pullCallable().call(pullRequest);
 
-			// Ack received messages.
 			if (pullResponse.getReceivedMessagesCount() > 0) {
-				List<String> ackIds = pullResponse.getReceivedMessagesList().stream()
-						.map(ReceivedMessage::getAckId)
-						.collect(Collectors.toList());
 
-				AcknowledgeRequest acknowledgeRequest = AcknowledgeRequest.newBuilder()
-						.setSubscription(pullRequest.getSubscription())
-						.addAllAckIds(ackIds)
-						.build();
-
-				subscriber.acknowledgeCallable().call(acknowledgeRequest);
+				pullResponse.getReceivedMessagesList().stream().forEach(
+						receivedMessage -> {
+							Map<String, String> attributesMap = new HashMap<>();
+							attributesMap.putAll(receivedMessage.getMessage().getAttributesMap());
+							attributesMap.put(GcpPubSubHeaders.ACK_ID, receivedMessage.getAckId());
+							attributesMap.put(GcpPubSubHeaders.SUBSCRIPTION,
+									pullRequest.getSubscription());
+							// TODO(joaomartins): Replace with toBuilder()?
+							pulledMessages.add(PubsubMessage.newBuilder()
```

Comment:
Returning a `PubsubMessage` instead of a `ReceivedMessage` was a conscious decision we took collectively, not to over-burden users with another "message" concept.
I'd prefer not to go back on that decision because I still think it's the best.

We have to rebuild the `PubsubMessage` because it is immutable. However, we might be able to use `PubsubMessage.toBuilder()`, which would allow us to do something similar to modifying a `PubsubMessage`.

Reply:
If we don't want to burden the users with yet another message type, maybe the right thing to do is to just return the SI `Message`?
@artembilan

Reply:
`PubSubTemplate` should be completely independent from SI.
We could invent a new `Message` object, but why doing so when there is already `PubsubMessage`? This has been the reasoning we've been following so far.

Reply:
I know we made the decision to go with `PubsubMessage` for simplicity in the past. However, I think we should revisit that decision in the light of these changes. `ReceivedMessage` is part of the official [Pub/Sub proto](https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub.proto). So, the concept should be familiar to anyone using Pub/Sub at a lower level of abstraction. It's also unlikely to change going forward.
Since the only additional thing `ReceivedMessage` contains is `ack_id`, we don't need to use it in the regular method that auto-acks, but a separate method that pull without acking should return `ReceivedMessage`.

Reply:
I agree about different method with a `ReceivedMessage` return type: `List<PubsubMessage> pullAutoAck()` and `List<ReceivedMessage> pull()`.

Reply:
After looking into this carefully, it seems that we can't escape from having to add the subscription to the message, since it isn't anywhere in `ReceivedMessage`.

Since we need to manipulate the message to add the subscription, I think we can keep going this way, also adding the ack ID and not having to introduce `ReceivedMessage`.